src/ray/rpc/worker/core_worker_client_pool.cc

```diff
@@ -33,32 +34,80 @@ std::function<void()> CoreWorkerClientPool::GetDefaultUnavailableTimeoutCallback
           raylet_client_factory = std::move(raylet_client_factory)]() {
     const NodeID node_id = NodeID::FromBinary(addr.raylet_id());
     const WorkerID worker_id = WorkerID::FromBinary(addr.worker_id());
-    RAY_CHECK(gcs_client->Nodes().IsSubscribedToNodeChange());
-    const rpc::GcsNodeInfo *node_info =
-        gcs_client->Nodes().Get(node_id, /*filter_dead_nodes=*/true);
-    if (node_info == nullptr) {
-      RAY_LOG(INFO).WithField(worker_id).WithField(node_id)
-          << "Disconnect core worker client since its node is dead";
-      worker_client_pool->Disconnect(worker_id);
+
+    auto check_worker_alive = [raylet_client_factory,
+                               worker_client_pool,
+                               worker_id,
+                               node_id](const rpc::GcsNodeInfo &node_info) {
+      auto raylet_client = raylet_client_factory(node_info.node_manager_address(),
+                                                 node_info.node_manager_port());
+      raylet_client->IsLocalWorkerDead(
+          worker_id,
+          [worker_client_pool, worker_id, node_id](const Status &status,
+                                                   rpc::IsLocalWorkerDeadReply &&reply) {
+            if (!status.ok()) {
+              // Will try again when unavailable timeout callback is retried.
+              RAY_LOG(INFO).WithField(worker_id).WithField(node_id)
+                  << "Failed to check if worker is dead on request to raylet";
+              return;
+            }
+            if (reply.is_dead()) {
+              RAY_LOG(INFO).WithField(worker_id).WithField(node_id)
+                  << "Disconnecting core worker client because the worker is dead";
+              worker_client_pool->Disconnect(worker_id);
+            }
+          });
+    };
+
+    auto gcs_check_node_alive =
+        [check_worker_alive, node_id, worker_id, worker_client_pool, gcs_client]() {
+          gcs_client->Nodes().AsyncGetAll(
+              [check_worker_alive = std::move(check_worker_alive),
+               worker_id,
+               node_id,
+               worker_client_pool](const Status &status,
+                                   std::vector<rpc::GcsNodeInfo> &&nodes) {
+                if (!status.ok()) {
+                  // Will try again when unavailable timeout callback is retried.
+                  RAY_LOG(INFO) << "Failed to get node info from GCS";
+                  return;
+                }
+                if (nodes.empty() || nodes[0].state() != rpc::GcsNodeInfo::ALIVE) {
+                  // The node is dead or the GCS has erased the info for this node. It
+                  // erases based on maximum_gcs_dead_node_cached_count. There's no way
+                  // for the a component to know about a remote node unless the gcs has
+                  // registered it.
```

Comment:
```suggestion
                  // The node is dead or GCS doesn't know about this node.
                  // There are two reasons why GCS doesn't know about a node: 1. The node hasn't register yet. 2. the GCS
                  // erased the dead node based on maximum_gcs_dead_node_cached_count. In our case, it must be 2 since there's no way
                  // for a component to know about a remote node id unless the gcs has
                  // registered it.
```